### PR TITLE
Rust agent developer guide documentation 

### PR DESCRIPTION
**Pull request by SirCipher**
Rendered: https://deploy-preview-207--scintillating-strudel-60ff56.netlify.app/server/rust/developer-guide/

### CODE REVIEW

File: src/_java-server-devguide/index.md [hunk 1]
```diff
@@ -1,13 +1,36 @@
...
-description: "Developer Guide to SwimOS"
-toc: false
 redirect_from:
-  - /developer-guide/
+  - /rust/developer-guide/introduction
+  - /rust/developer-guide/introduction.html
 ---
 
-This ordered guide will introduce you step by step to SwimOS, starting with the fundamentals and working up to some more advanced features.
+Welcome to the SwimOS developer's guide, an introductory guide to the SwimOS framework. With SwimOS, you can dive into building applications without getting bogged down by intricate infrastructure details – it takes care of the 'plumbing' so you can concentrate on crafting robust application logic.
```
Comment (horned-sphere) [inline src/_java-server-devguide/index.md:12]: Bogged down is too colloquial.

Comment (horned-sphere) [reply src/_java-server-devguide/index.md:12]: "Plumbing" is a big vague. I know what you mean, but the average reader might not.

File: src/_java-server-devguide/index.md [hunk 2]
```diff
@@ -1,13 +1,36 @@
...
 
-{% include docs-listing.html %}
+In this guide, we will walk through creating a simple SwimOS server that stores the state of a single i32 value in an agent and then interact with the server using a client application. The guide will introduce some core SwimOS concepts: agents and lanes. Our goal is to equip you with the knowledge and skills to harness the full potential of SwimOS for developing real-time applications. While we'll cover the 'how' of building streaming apps, if you're interested in understanding the 'why' behind it, you may find our reference documentation more insightful.
```
Comment (horned-sphere) [inline src/_java-server-devguide/index.md:14]: I would say "32bit signed integer" in prose sections rather than using the type name. Where you do use a type name, you should monospace format it.

Comment (horned-sphere) [reply src/_java-server-devguide/index.md:14]: Where you mention introducing core concepts, you should link out to the documentation that describes them if it's not going to happen here.

Comment (horned-sphere) [reply src/_java-server-devguide/index.md:14]: I'd rephrase the 'how' and 'why' bit to be more normal sentences. It's kinder to non-native speakers.

File: src/_java-server-devguide/index.md [hunk 3]
```diff
@@ -1,13 +1,36 @@
...
+
+SwimOS Rust requires a minimum Rust version of `{{ site.data.rust.rustc-version }}`. You can check your current Rust version using:
+
+```shell
+$ rustc --version
+```
+
+Which should return an output of something like `rustc 1.76.0 (07dca489a 2024-02-04)`.
+
+If you do not have a version that matches this requirement you can update it using:
```
Comment (horned-sphere) [inline src/_java-server-devguide/index.md:32]: 'matches' => 'satisfies' or 'meets'

File: src/_rust-server-devguide/1-introduction.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+short-title: Introduction
+description: "Introduction to the SwimOS developer guide."
+group: Introduction
+layout: documentation
+redirect_from:
+  - /rust/developer-guide/introduction
+  - /rust/developer-guide/introduction.html
+---
+
+Welcome to the SwimOS developer's guide, an introductory guide to the SwimOS framework. With SwimOS, you can dive into building applications without getting bogged down by intricate infrastructure details – it takes care of the 'plumbing' so you can concentrate on crafting robust application logic.
```
Comment (horned-sphere) [inline src/_rust-server-devguide/1-introduction.md:12]: This just seems to rephrase the content from the index. Do you really want to repeat yourself? If this is intended, they should probably be indentical and possibly be imported from another file.

Comment (horned-sphere) [reply src/_rust-server-devguide/1-introduction.md:12]: My comments to the passage on the index apply here too if it stays.

File: src/_rust-server-devguide/2-getting-started.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+short-title: Getting Started
+description: "Getting Started using SwimOS."
+group: Getting Started
+layout: documentation
+redirect_from:
+  - /rust/developer-guide/getting-started
+  - /rust/developer-guide/getting-started.html
+---
+
+Let's get started on the SwimOS journey.
```
Comment (horned-sphere) [inline src/_rust-server-devguide/2-getting-started.md:12]: I really don't like "journeys" but that's just personal taste ;).

File: src/_rust-server-devguide/2-getting-started.md [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+In this chapter we will discuss project setup and create the most basic SwimOS 'Hello, world!' server and extend it to incorporate some SwimOS core concepts, agents and lanes.
```
Comment (horned-sphere) [inline src/_rust-server-devguide/2-getting-started.md:13]: I _think_ it should be a colon rather than a comma.

File: src/_rust-server-devguide/2.1-project-setup.md [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+```shell
+$ cargo new tutorial_server --bin
+```
+
+Then we need to declare our dependencies. Let's start by adding the following to our `tutorial_server/Cargo.toml`:
+
+```toml
+[dependencies]
+swimos = "{{ site.data.rust.swimos-version }}"
+tokio = { "{{ site.data.rust.tokio-version }}", features = ["full"] }
```
Comment (horned-sphere) [inline src/_rust-server-devguide/2.1-project-setup.md:23]: Might be better to only add the features you actually use (it shouldn't be many for an example app).

File: src/_rust-server-devguide/2.2-simple-agent.md [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+redirect_from:
+  - /rust/developer-guide/simple-agent
+  - /rust/developer-guide/simple-agent.html
+---
+
+This section of the guide will walk through creating a simple SwimOS server application that stores the state of a single i32 value in an agent and interacting with it using a client application.
+
+# Agents
+
+In SwimOS, a Web Agent is a structural representation of an entity that lives amongst your streaming data. It continuously consumes updates of synced state and for each update, it computes new state that automatically gets synced to listeners. These Web Agents contain Lanes which store your state and provide access to it through an address known as a lane URI.
```
Comment (horned-sphere) [inline src/_rust-server-devguide/2.2-simple-agent.md:16]: I don't think this paragraph would be very clear to someone who doesn't know what Swim really does.

File: src/_rust-server-devguide/2.2-simple-agent.md [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+
+In SwimOS, there are two points of addressability: an agent URI and a lane URI. These are loosly analogous to a database row and column; where the row is the item of data (Web Agent) and the column is the field (lane). Agent URIs may be static (where only one will exist) or dynamic (potentially unbounded); these are covered in more detail later in this guide. This fine-grained addressability allows you to update state and receive updates for only the data that you are interested in.
```
Comment (horned-sphere) [inline src/_rust-server-devguide/2.2-simple-agent.md:18]: I don't think the row/column comparison is good. I'd probably use the agent model (or object orentiation if you want something less accurate but more broadly understood).

Comment (horned-sphere) [reply src/_rust-server-devguide/2.2-simple-agent.md:18]: I think saying an address has two parts would be clearer than "two points of addressability". You should probably explain how the agent URIs work here too (including the possibility of having parameterized agents).

File: src/_rust-server-devguide/2.2-simple-agent.md [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+}
+```
+
+Lets go over the block of code above.
+
+```rust
+#[derive(AgentLaneModel)]
+```
+
+The `#[derive(AgentLaneModel)]` macro derives a descriptor of the agent so the agent's runtime is aware of its lane URIs, whether they are transient, how to route messages to each lane and the agent's constructor; a Web Agent may only contain SwimOS core types such as the `ValueLane` above.
```
Comment (horned-sphere) [inline src/_rust-server-devguide/2.2-simple-agent.md:46]: You should be consistent between calling them "agent" or "web agent".

File: src/_rust-server-devguide/2.2-simple-agent.md [hunk 11]
```diff
@@ -0,0 +1,432 @@
...
+
+```rust
+#[projections]
+```
+
+`#[projections]` declares a mapping from the agent to each lane within the agent; for the example above, this is: `fn(&ExampleAgent) -> &ValueLane<i32>`. These projections are required when an agent or lane handler is used to perform an action such as updating the state of a lane; handlers are covered later in this guide.
```
Comment (horned-sphere) [inline src/_rust-server-devguide/2.2-simple-agent.md:52]: I'd probably defer explaining this until you're actually writing a handler.

File: src/_rust-server-devguide/2.2-simple-agent.md [hunk 12]
```diff
@@ -0,0 +1,432 @@
...
+#[derive(Clone)]
+pub struct ExampleLifecycle;
+
+#[lifecycle(ExampleAgent)]
+impl ExampleLifecycle {
+
+}
+```
+
+Every agent has a lifecycle associated with it and here one is declared and `ExampleLifecycle`'s implementation is bound to the `ExampleAgent`. While an empty lifecycle has been declared, it is possible to add functions which will be invoked throughout the lifecycle of the agent, such as when the agent starts, a lane receives an update, as well as many other events; one is added later in this guide.
```
Comment (horned-sphere) [inline src/_rust-server-devguide/2.2-simple-agent.md:70]: Explain that the agent definition and the lifecycles are decoupled.

File: src/_rust-server-devguide/2.2-simple-agent.md [hunk 13]
```diff
@@ -0,0 +1,432 @@
...
+
+- Static: such as `"/example"`. If a static path is used, only one agent will ever exist at that address.
+- Dynamic: such as `"/example/:id"`. For each unique address, a new agent will be instantiated.
+
+For this guide, a dynamic route of `"/example/:id"` will be used.
+
+Append the following to `tutorial_server/src/main.rs`:
+
+```rust
+use std::{error::Error, time::Duration};
```
Comment (horned-sphere) [inline src/_rust-server-devguide/2.2-simple-agent.md:84]: I'd do this in two chunks. First explain how to set up the server (with no agents). Then a section or registering the agent routes.

File: src/_rust-server-devguide/2.2-simple-agent.md [hunk 14]
```diff
@@ -0,0 +1,432 @@
...
+use swimos::{
+    agent::agent_model::AgentModel,
+    route::RoutePattern,
+    server::{ServerBuilder, ServerHandle, Server},
+};
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn Error + Send + Sync>> {
+    // Create a dynamic route for our agents.
```
Comment (horned-sphere) [inline src/_rust-server-devguide/2.2-simple-agent.md:93]: Maybe thin down the comments and do more of the explanation in prose. It's a bit too much code vomit for a step by step tutorial.

File: src/_rust-server-devguide/2.2-simple-agent.md [hunk 15]
```diff
@@ -0,0 +1,432 @@
...
+    // Join on the server and ctrl+c futures.
+    let (_, result) = tokio::join!(shutdown, task);
+
+    result?;
+    println!("Server stopped successfully.");
+    Ok(())
+}
+
+// Utility function for awaiting a stop signal in the terminal.
+async fn manage_handle(mut handle: ServerHandle) {
```
Comment (horned-sphere) [inline src/_rust-server-devguide/2.2-simple-agent.md:131]: This should be extracted as a utiluty somwhere so it doesn't clutter up all the examples.

File: src/_rust-server-devguide/2.2-simple-agent.md [hunk 16]
```diff
@@ -0,0 +1,432 @@
...
+    let agent = AgentModel::new(ExampleAgent::default, ExampleLifecycle.into_lifecycle());
+
+    // Create a server builder.
+    let server = ServerBuilder::with_plane_name("Plane")
+        // Bind to port 8080
+        .set_bind_addr("127.0.0.1:8080".parse().unwrap())
+        // For this guide, ensure agents timeout fairly quickly.
+        // An agent will timeout after they have received no new updates
+        // for this configured period of time.
+        .update_config(|config| {
```
Comment (horned-sphere) [inline src/_rust-server-devguide/2.2-simple-agent.md:106]: Probably don't fiddle with the default config for a tutorial to reduce noise.

File: src/_rust-server-devguide/2.2-simple-agent.md [hunk 17]
```diff
@@ -0,0 +1,432 @@
...
+    // Create a dynamic route for our agents.
+    let route = RoutePattern::parse_str("/example/:id")?;
+    // Create an agent model which contains the factory for creating the agent as well
+    // as the lifecycle which will be run.
+    let agent = AgentModel::new(ExampleAgent::default, ExampleLifecycle.into_lifecycle());
+
+    // Create a server builder.
+    let server = ServerBuilder::with_plane_name("Plane")
+        // Bind to port 8080
+        .set_bind_addr("127.0.0.1:8080".parse().unwrap())
```
Comment (horned-sphere) [inline src/_rust-server-devguide/2.2-simple-agent.md:102]: Do we have a default bind address? If so, omit this for the tutorial (and add a footnote that there is a default).

File: src/_rust-server-devguide/2.2-simple-agent.md [hunk 18]
```diff
@@ -0,0 +1,432 @@
...
+You can now run the SwimOS server application:
+
+```shell
+$ cd tutorial_server
+$ cargo run
+```
+
+At this point, nothing will be seen in the console output as the server is listening for events and none are being produced.
+
+# Interacting with the server
```
Comment (horned-sphere) [inline src/_rust-server-devguide/2.2-simple-agent.md:150]: I'd probably do multiple pages with navigation. This could get long winded.

Comment (horned-sphere) [reply src/_rust-server-devguide/2.2-simple-agent.md:150]: Initially, maybe get them to use the command line clinet instead. It's less code overhead to get something that works. Then introduce the clinet library separately.

File: src/_rust-server-devguide/2.2-simple-agent.md [hunk 19]
```diff
@@ -0,0 +1,432 @@
...
+    );
+
+    let lifecycle = BasicValueDownlinkLifecycle::<usize>::default()
+        // Register an event handler that is invoked when the downlink connects to the agent.
+        .on_linked_blocking(|| println!("Downlink linked"))
+        // Register an event handler that is invoked when the downlink synchronises its state.
+        // with the agent.
+        .on_synced_blocking(|value| println!("Downlink synced with: {value:?}"))
+        // Register an event handler that is invoked when the downlink receives an event.
+        .on_event_blocking(|value| println!("Downlink event: {value:?}"));
```
Comment (horned-sphere) [inline src/_rust-server-devguide/2.2-simple-agent.md:202]: Unnecessary nouse for a tutorial. Just go for `on_event`.

File: src/_rust-server-devguide/2.2-simple-agent.md [hunk 20]
```diff
@@ -0,0 +1,432 @@
...
+    // Build a path the downlink.
+    let state_path = RemotePath::new(
+        // The host address
+        "ws://0.0.0.0:8080",
+        // You can provide any agent URI that matches the pattern
+        // "/example/:id"
+        "/example/1",
+        // This is the URI of the ValueLane<i32> in our ExampleAgent
+        "state",
+    );
```
Comment (horned-sphere) [inline src/_rust-server-devguide/2.2-simple-agent.md:193]: Again, thin comments and move into prose.

File: src/_rust-server-devguide/2.2-simple-agent.md [hunk 21]
```diff
@@ -0,0 +1,432 @@
...
+        // Register an event handler that is invoked when the downlink synchronises its state.
+        // with the agent.
+        .on_synced_blocking(|value| println!("Downlink synced with: {value:?}"))
+        // Register an event handler that is invoked when the downlink receives an event.
+        .on_event_blocking(|value| println!("Downlink event: {value:?}"));
+
+    // Build our downlink.
+    //
+    // This operation may fail if there is a connection issue.
+    let state_downlink = handle
```
Comment (horned-sphere) [inline src/_rust-server-devguide/2.2-simple-agent.md:207]: Explain that the downlink runs on its own task, otherwise people might be surprised that the set calls below do something.

File: src/_rust-server-devguide/2.2-simple-agent.md [hunk 22]
```diff
@@ -0,0 +1,432 @@
...
+        .open()
+        .await
+        .expect("Failed to open downlink");
+
+    for i in 0..10 {
+        // Update the lane's state.
+        state_downlink.set(i).await.expect("Failed to set downlink state");
+    }
+
+    tokio::signal::ctrl_c()
```
Comment (horned-sphere) [inline src/_rust-server-devguide/2.2-simple-agent.md:220]: Explain what would happen if this wasn't here.

File: src/_rust-server-devguide/2.2-simple-agent.md [hunk 23]
```diff
@@ -0,0 +1,432 @@
...
+Downlink event: 3
+Downlink event: 4
+Downlink event: 5
+Downlink event: 6
+Downlink event: 7
+Downlink event: 8
+Downlink event: 9
+```
+
+Inspecting these events, you can observe an execution flow of `linked -> synced -> events (0 -> 9)`.
```
Comment (horned-sphere) [inline src/_rust-server-devguide/2.2-simple-agent.md:255]: All of the lifecycle stuff should probably be in a section that describes what a downlink is in the abstract (and linked from here).

File: src/_rust-server-devguide/2.2-simple-agent.md [hunk 24]
```diff
@@ -0,0 +1,432 @@
...
+
+# Agent Lifecycle
+
+Agent lifecycles provide the ability to build powerful, stateful, streaming data applications and they are where you will spend most of your time when writing SwimOS server applications. Inside an agent lifecycle implementation, a number of event handlers may be registered which are invoked at various stages within the lifecycle. There are three types of lifecycle event handlers:
+
+- The `on_start` handler: invoked exactly once when the agent has started. This handler may be useful if you need an agent to start polling an API once the agent has started.
+- The `on_stop` handler: invoked exactly once just before the agent stops. This handler may be useful to clear up some resources before an agent stops.
+- Lane-specific event handlers which are invoked when a lane receives an event. For example, for a `ValueLane` you may register a handler that is invoked after the state of the lane has changed. These handlers allow your agent to _react_ to the events that it receives and to trigger updates elsewhere.
```
Comment (horned-sphere) [inline src/_rust-server-devguide/2.2-simple-agent.md:267]: I'd splt the two parts entirely, `on_start` and `on_stop` are agent lifecycle events and the rest are lane lifecycle events. Given them their own sections.

File: src/_rust-server-devguide/2.2-simple-agent.md [hunk 25]
```diff
@@ -0,0 +1,432 @@
...
+    pub fn on_stop(
+        &self,
+        context: HandlerContext<ExampleAgent>,
+    ) -> impl EventHandler<ExampleAgent> {
+        context.effect(|| println!("Stopping agent."))
+    }
+}
+```
+
+Each lifecycle event handler accepts a `HandlerContext` (which aids in the creation of `EventHandlers`) and returns an `EventHandler` which is scoped to the agent that the lifecycle is bound to; this allows the handler to interact with the lanes within the agent. An `EventHandler` models an action that happens within an agent after an event is triggered; such as an agent starting or an event was received. They may not return a value but may trigger an event to happen elsewhere within an agent, such as updating the state of a lane or running an async task; these handlers are created using the provided `HandlerContext`.
```
Comment (horned-sphere) [inline src/_rust-server-devguide/2.2-simple-agent.md:299]: Maybe just say that it's a factory used to create handlers.

File: src/_java-server-devguide/index.md (hunk 1 above)
Comment (SirCipher) [reply src/_java-server-devguide/index.md:12]: I think the branches may have diverged here. I didn't actually write this - I've only touched the rust directories.

File: src/_rust-server-devguide/2.2-simple-agent.md (hunk 17 above)
Comment (SirCipher) [reply src/_rust-server-devguide/2.2-simple-agent.md:102]: Default is 0 at the moment

File: src/_rust-server-devguide/2.2-simple-agent.md (hunk 18 above)
Comment (SirCipher) [reply src/_rust-server-devguide/2.2-simple-agent.md:150]: I think getting users to use the CLI would be even more longwinded as it would require cloning the repo, building and then using it - given that the client will be introduced anyway. What do you think?
